File: bindings/dotnet/src/Ccharts/Chart.cs
using System.Runtime.InteropServices;

namespace Ccharts;

/// <summary>
/// A parsed OHLC dataset that can be rendered as a line or candlestick chart.
/// </summary>
/// <remarks>
/// <para>
/// Charts are immutable once built and the native library holds no mutable
/// global state, so an instance is safe to render from several threads.
/// </para>
/// <para>
/// <see cref="Dispose"/> releases the native memory; a finalizer covers a
/// forgotten call.
/// </para>
/// <example>
/// <code>
/// using var chart = Chart.FromArrays(open, high, low, close, epochSeconds);
/// Console.Write(chart.Line(new ChartOptions { Width = 60, ShowPrices = true }));
/// </code>
/// </example>
/// </remarks>
public sealed class Chart : IDisposable
{
    private static readonly ChartOptions DefaultOptions = new();

    private IntPtr _handle;
    private readonly int _length;

    private Chart(IntPtr handle)
    {
        _handle = handle;
        _length = NativeMethods.DataLength(handle);
    }

    /// <summary>
    /// Builds a chart from four equal-length price columns.
    /// </summary>
    /// <param name="open">Opening prices.</param>
    /// <param name="high">High prices.</param>
    /// <param name="low">Low prices.</param>
    /// <param name="close">Closing prices.</param>
    /// <param name="ts">
    /// Epoch seconds, or an empty span when the candles have no timestamps.
    /// </param>
    /// <exception cref="CchartsException">
    /// The columns were empty, of differing lengths, or held NaN/infinity.
    /// </exception>
    public static Chart FromArrays(
        ReadOnlySpan<double> open, ReadOnlySpan<double> high,
        ReadOnlySpan<double> low, ReadOnlySpan<double> close,
        ReadOnlySpan<long> ts = default)
    {
        NativeLibraryResolver.EnsureInstalled();

        var n = open.Length;
        if (n == 0)
        {
            throw new CchartsException(CchartsStatus.InvalidArgument,
                "need at least one candle");
        }
        if (high.Length != n || low.Length != n || close.Length != n)
        {
            throw new CchartsException(CchartsStatus.InvalidArgument,
                "open, high, low and close must have the same length");
        }
        if (!ts.IsEmpty && ts.Length != n)
        {
            throw new CchartsException(CchartsStatus.InvalidArgument,
                "ts must have the same length as the price columns");
        }

        // An empty span marshals to a null pointer, which the ABI reads as
        // "no timestamps".
        var status = NativeMethods.FromArrays(open, high, low, close, ts, n, out var handle);
        CchartsException.ThrowIfError(status);
        return new Chart(handle);
    }

    /// <summary>
    /// Builds a chart from the fixed-schema JSON document: an array of objects
    /// with <c>ts</c>, <c>open</c>, <c>high</c>, <c>low</c> and <c>close</c>.
    /// </summary>
    /// <exception cref="CchartsException">The document could not be parsed.</exception>
    public static Chart FromJson(string json)
    {
        ArgumentNullException.ThrowIfNull(json);
        NativeLibraryResolver.EnsureInstalled();

        var utf8 = Marshal.StringToCoTaskMemUTF8(json);
        try
        {
            CchartsException.ThrowIfError(NativeMethods.ParseJson(utf8, out var handle));
            return new Chart(handle);
        }
        finally
        {
            Marshal.FreeCoTaskMem(utf8);
        }
    }

    /// <summary>
    /// Builds a chart from CSV rows of <c>open,high,low,close[,timestamp]</c>.
    /// Blank lines are skipped.
    /// </summary>
    public static Chart FromCsv(string csv, char valueSeparator = ',', char lineSeparator = '\n')
    {
        ArgumentNullException.ThrowIfNull(csv);
        if (valueSeparator > 127 || lineSeparator > 127 ||
            valueSeparator == '\0' || lineSeparator == '\0')
        {
            throw new CchartsException(CchartsStatus.InvalidArgument,
                "separators must be non-NUL ASCII characters");
        }
        NativeLibraryResolver.EnsureInstalled();

        var utf8 = Marshal.StringToCoTaskMemUTF8(csv);
        try
        {
            var status = NativeMethods.ParseCsv(utf8, (byte)valueSeparator,
                (byte)lineSeparator, out var handle);
            CchartsException.ThrowIfError(status);
            return new Chart(handle);
        }
        finally
        {
            Marshal.FreeCoTaskMem(utf8);
        }
    }

    /// <summary>Number of candles in the dataset.</summary>
    public int Length => _length;

    /// <summary>
    /// Renders a pie/donut chart from the given slices. A pie has no OHLC
    /// dataset, so this is a static method taking the slices directly.
    /// </summary>
    /// <param name="slices">The slices, in the order they are drawn.</param>
    /// <param name="options">Pie options, or <c>null</c> for the defaults.</param>
    /// <returns>The chart as a printable string.</returns>
    /// <exception cref="CchartsException">
    /// The slices were empty, held NaN/infinity, or the dimensions were out of
    /// range.
    /// </exception>
    public static string Pie(IReadOnlyList<PieSlice> slices, PieOptions? options = null)
    {
        ArgumentNullException.ThrowIfNull(slices);
        NativeLibraryResolver.EnsureInstalled();

        options ??= new PieOptions();
        if (slices.Count == 0)
        {
            throw new CchartsException(CchartsStatus.InvalidArgument,
                "need at least one slice");
        }

        var native = new NativePieSlice[slices.Count];
        var labels = new IntPtr[slices.Count];
        try
        {
            for (var i = 0; i < slices.Count; i++)
            {
                labels[i] = slices[i].Label is null ? IntPtr.Zero
                    : Marshal.StringToCoTaskMemUTF8(slices[i].Label);
                native[i] = new NativePieSlice { Label = labels[i], Value = slices[i].Value };
            }

            IntPtr[]? colors = null;
            if (options.Colors is { Count: > 0 })
            {
                colors = new IntPtr[options.Colors.Count];
                for (var i = 0; i < options.Colors.Count; i++)
                {
                    var color = options.Colors[i];
                    colors[i] = string.IsNullOrEmpty(color) ? IntPtr.Zero
                        : Marshal.StringToCoTaskMemUTF8(color);
                }
            }

            try
            {
                var centerText = string.IsNullOrEmpty(options.CenterText) ? IntPtr.Zero
                    : Marshal.StringToCoTaskMemUTF8(options.CenterText);
                try
                {
                    var status = NativeMethods.PieFromSlices(
                        native, slices.Count, options.Width, options.Height,
                        options.Donut ? 1 : 0, colors, colors?.Length ?? 0,
                        options.ShowLegend ? 1 : 0, options.ShowPct ? 1 : 0,
                        options.SliceGap, options.InnerRadiusRatio,
                        options.LegendFormat, options.StartAngle,
                        options.CounterClockwise ? 1 : 0, centerText,
                        out var chart, out var length);
                    CchartsException.ThrowIfError(status);

                    try
                    {
                        return Marshal.PtrToStringUTF8(chart, checked((int)length)) ?? string.Empty;
                    }
                    finally
                    {
                        NativeMethods.StringFree(chart);
                    }
                }
                finally
                {
                    if (centerText != IntPtr.Zero)
                    {
                        Marshal.FreeCoTaskMem(centerText);
                    }
                }
            }
            finally
            {
                if (colors is not null)
                {
                    foreach (var color in colors)
                    {
                        if (color != IntPtr.Zero)
                        {
                            Marshal.FreeCoTaskMem(color);
                        }
                    }
                }
            }
        }
        finally
        {
            foreach (var label in labels)
            {
                if (label != IntPtr.Zero)
                {
                    Marshal.FreeCoTaskMem(label);
                }
            }
        }
    }

    /// <summary>
    /// Renders a histogram of the given scalar samples. A histogram has no
    /// OHLC dataset, so this is a static method taking the raw sample values.
    /// </summary>
    /// <param name="samples">The scalar values to bin.</param>
    /// <param name="options">Histogram options, or <c>null</c> for the defaults.</param>
    /// <returns>The chart as a printable string.</returns>
    /// <exception cref="CchartsException">
    /// The samples were empty or held NaN/infinity, or the dimensions were out
    /// of range.
    /// </exception>
    public static string Histogram(IReadOnlyList<double> samples, HistogramOptions? options = null)
    {
        ArgumentNullException.ThrowIfNull(samples);
        NativeLibraryResolver.EnsureInstalled();

        options ??= new HistogramOptions();
        if (samples.Count == 0)
        {
            throw new CchartsException(CchartsStatus.InvalidArgument,
                "need at least one sample");
        }

        var colors = new IntPtr[2];
        try
        {
            // An empty C string means "emit no escape at all", which is
            // different from a null pointer (use the default color).
            colors[0] = options.Plain ? Empty() : Utf8(options.Color);
            colors[1] = options.Plain ? Empty() : Utf8(options.BackgroundColor);

            var settings = new NativeHistSettings
            {
                RiseColor = colors[0],
                BackgroundColor = colors[1],
                BinCount = options.BinCount,
                // null is the NaN "auto" sentinel: use the data min/max.
                MinValue = options.MinValue ?? double.NaN,
                MaxValue = options.MaxValue ?? double.NaN,
                ShowBins = options.ShowBins ? 1 : 0,
                ShowPrices = options.ShowPrices ? 1 : 0,
            };

            var status = NativeMethods.Hist(samples.ToArray(), samples.Count,
                options.Width, options.Height, in settings, out var chart, out var length);
            CchartsException.ThrowIfError(status);

            try
            {
                return Marshal.PtrToStringUTF8(chart, checked((int)length));
            }
            finally
            {
                NativeMethods.StringFree(chart);
            }
        }
        finally
        {
            foreach (var color in colors)
            {
                if (color != IntPtr.Zero)
                {
                    Marshal.FreeCoTaskMem(color);
                }
            }
        }
    }

    /// <summary>Version of the underlying C library.</summary>
    public static string Version
    {
        get
        {
            NativeLibraryResolver.EnsureInstalled();
            return Marshal.PtrToStringUTF8(NativeMethods.Version()) ?? string.Empty;
        }
    }

    /// <summary>Largest width or height in cells (CC_MAX_DIM).</summary>
    public static int MaxDim
    {
        get
        {
            NativeLibraryResolver.EnsureInstalled();
            return NativeMethods.MaxDim();
        }
    }

    /// <summary>Largest number of cells in a chart (CC_MAX_CELLS).</summary>
    public static int MaxCells
    {
        get
        {
            NativeLibraryResolver.EnsureInstalled();
            return NativeMethods.MaxCells();
        }
    }

    /// <summary>Renders a line chart of the closing prices.</summary>
    /// <exception cref="CchartsException">The dimensions were out of range.</exception>
    public string Line(ChartOptions? options = null) => Render(line: true, options);

    /// <summary>Renders a candlestick chart.</summary>
    /// <exception cref="CchartsException">The dimensions were out of range.</exception>
    public string Candle(ChartOptions? options = null) => Render(line: false, options);

    private string Render(bool line, ChartOptions? options)
    {
        ObjectDisposedException.ThrowIf(_handle == IntPtr.Zero, this);
        options ??= DefaultOptions;

        var colors = new IntPtr[4];
        try
        {
            // An empty C string means "emit no escape at all", which is
            // different from a null pointer (use the default color).
            colors[0] = options.Plain ? Empty() : Utf8(options.RiseColor);
            colors[1] = options.Plain ? Empty() : Utf8(options.FallColor);
            colors[2] = options.Plain ? Empty() : Utf8(options.BackgroundColor);
            colors[3] = options.Plain ? Empty() : Utf8(options.AreaColor);

            var settings = new NativeSettings
            {
                RiseColor = colors[0],
                FallColor = colors[1],
                BackgroundColor = colors[2],
                AreaColor = colors[3],
                SingleColor = options.SingleColor ? 1 : 0,
                ShowPrices = options.ShowPrices ? 1 : 0,
                ShowTimes = options.ShowTimes ? 1 : 0,
            };

            var status = line
                ? NativeMethods.Line(_handle, options.Width, options.Height, in settings,
                    out var chart, out var length)
                : NativeMethods.Candle(_handle, options.Width, options.Height, in settings,
                    out chart, out length);
            CchartsException.ThrowIfError(status);

            try
            {
                return Marshal.PtrToStringUTF8(chart, checked((int)length));
            }
            finally
            {
                NativeMethods.StringFree(chart);
            }
        }
        finally
        {
            foreach (var color in colors)
            {
                if (color != IntPtr.Zero)
                {
                    Marshal.FreeCoTaskMem(color);
                }
            }
        }
    }

    private static IntPtr Utf8(string? value) =>
        string.IsNullOrEmpty(value) ? IntPtr.Zero : Marshal.StringToCoTaskMemUTF8(value);

    private static IntPtr Empty() => Marshal.StringToCoTaskMemUTF8(string.Empty);

    /// <summary>Releases the native dataset. Safe to call more than once.</summary>
    public void Dispose()
    {
        var handle = Interlocked.Exchange(ref _handle, IntPtr.Zero);
        if (handle != IntPtr.Zero)
        {
            NativeMethods.DataFree(handle);
        }
        GC.SuppressFinalize(this);
    }

    /// <summary>Releases the native dataset if <see cref="Dispose"/> was not called.</summary>
    ~Chart() => Dispose();
}

File: bindings/ruby/lib/ccharts/chart.rb
# frozen_string_literal: true

module Ccharts
  # High-level chart API. A dataset built with {Chart.from_arrays},
  # {Chart.from_json} or {Chart.from_csv} is immutable; the pie / histogram /
  # sparkline / bar / stacked_bar / heatmap / boxplot renderers take their data
  # directly (they have no OHLC dataset, exactly like the C ABI they wrap).
  class Chart
    # ------------------------------------------------------------------
    # Building a dataset
    # ------------------------------------------------------------------

    # Build a dataset from four equal-length price columns and optional epoch
    # seconds. `ts` may be nil (all timestamps unknown).
    def self.from_arrays(open:, high:, low:, close:, ts: nil)
      n = open.length
      raise Error, "need at least one candle" if n == 0
      unless [high, low, close].all? { |a| a.length == n }
        raise Error, "open, high, low and close must have the same length"
      end
      if ts && ts.length != n
        raise Error, "ts must have the same length as the price columns"
      end

      open_ptr = FFI.pack_doubles(open)
      high_ptr = FFI.pack_doubles(high)
      low_ptr = FFI.pack_doubles(low)
      close_ptr = FFI.pack_doubles(close)
      ts_ptr = ts ? FFI.pack_i64(ts) : Fiddle::NULL

      out = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
      status = FFI::FROM_ARRAYS.call(open_ptr, high_ptr, low_ptr, close_ptr, ts_ptr, n, out)
      _from_data_status(status, out)
    end

    # Build a dataset from the fixed-schema JSON document accepted by the
    # C layer (an array of {ts, open, high, low, close} objects).
    def self.from_json(text)
      json = FFI.cstr(text)
      out = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
      status = FFI::PARSE_JSON.call(json, out)
      _from_data_status(status, out)
    end

    # Build a dataset from CSV rows of `open,high,low,close[,timestamp]`.
    def self.from_csv(text, value_separator: ",", line_separator: "\n")
      csv = FFI.cstr(text)
      vs = value_separator.ord
      ls = line_separator.ord
      raise Error, "separators must not be NUL" if vs == 0 || ls == 0
      out = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
      status = FFI::PARSE_CSV.call(csv, vs, ls, out)
      _from_data_status(status, out)
    end

    def self._from_data_status(status, out)
      Ccharts.raise_on(status)
      addr = FFI.read_ptr_int(out, Fiddle::SIZEOF_VOIDP)
      raise Error, "no dataset handle returned" if addr.zero?
      _from_handle(Fiddle::Pointer.new(addr))
    end

    # ------------------------------------------------------------------
    # Rendering an OHLC dataset (line / candle)
    # ------------------------------------------------------------------

    def line(width, height, settings = nil)
      _render(FFI::LINE, width, height, settings || Settings::ChartSettings.new)
    end

    def candle(width, height, settings = nil)
      _render(FFI::CANDLE, width, height, settings || Settings::ChartSettings.new)
    end

    # Number of candles in this dataset.
    def size
      FFI::DATA_LEN.call(@handle)
    end
    alias_method :length, :size

    # ------------------------------------------------------------------
    # Pie / histogram / sparkline / bar / stack / heat / box renderers
    # ------------------------------------------------------------------

    def self.pie(slices, width, height, settings = nil)
      raise Error, "need at least one slice" if slices.empty?
      settings ||= Settings::PieSettings.new

      keep = []
      rows = slices.map do |sl|
        label = sl[:label] || sl["label"]
        value = (sl[:value] || sl["value"] || 0.0).to_f
        lp = FFI.cstr(label)
        keep << lp
        [[0, 8, "Q", lp.to_i], [8, 8, "d", value]] # label @0, value @8
      end
      slices_ptr = FFI.contiguous_structs(16, rows) # ccharts_pie_slice = 16 B
      keep << slices_ptr

      colors_ptr_addr = settings.colors_ptr(keep).to_i
      center_ptr = settings.center_text_ptr(keep)
      donut, show_legend, show_pct, slice_gap, inner_radius_ratio,
        legend_format, start_angle, counter_clockwise = settings._scalars

      # `_scalars` already returns the flags as C ints 0/1 — pass them straight
      # through. Do NOT write `x ? 1 : 0` here: in Ruby `0` is truthy, so a
      # zeroed flag would be coerced to 1 (pct/ccw always on → mirrored pie).
      _read_string(
        FFI::PIE, slices_ptr, slices.length, width, height,
        donut, colors_ptr_addr, settings.colors_count,
        show_legend, show_pct,
        slice_gap, inner_radius_ratio, legend_format,
        start_angle, counter_clockwise, center_ptr
      )
    end

    def self.histogram(samples, width, height, settings = nil)
      raise Error, "need at least one sample" if samples.empty?
      settings ||= Settings::HistSettings.new
      ptr = FFI.pack_doubles(samples)
      s, keep = settings.to_ffi
      _read_string(FFI::HIST, ptr, samples.length, width, height, s.to_ptr)
    end

    def self.sparkline(samples, width, height, settings = nil)
      raise Error, "need at least one sample" if samples.empty?
      settings ||= Settings::SparkSettings.new
      ptr = FFI.pack_doubles(samples)
      s, keep = settings.to_ffi
      _read_string(FFI::SPARK, ptr, samples.length, width, height, s.to_ptr)
    end

    def self.bar(labels, values, width, height, settings = nil)
      raise Error, "need at least one bar" if labels.empty?
      raise Error, "labels and values must have the same length" unless labels.length == values.length
      settings ||= Settings::BarSettings.new

      keep = []
      rows = labels.zip(values).map do |lbl, val|
        lp = FFI.cstr(lbl)
        keep << lp
        [[0, 8, "Q", lp.to_i], [8, 8, "d", val.to_f]] # label @0, value @8
      end
      items_ptr = FFI.contiguous_structs(16, rows) # ccharts_bar_slice = 16 B
      keep << items_ptr
      s, skeep = settings.to_ffi
      keep.concat(skeep)
      _read_string(FFI::BAR, items_ptr, labels.length, width, height, s.to_ptr)
    end

    # series: array of {name:, values:} hashes sharing one category count.
    def self.stacked_bar(series, width, height, settings = nil)
      raise Error, "need at least one series" if series.empty?
      cats = series[0][:values]&.length || series[0]["values"].length
      raise Error, "series values must not be empty" if cats == 0
      series.each do |sv|
        vs = sv[:values] || sv["values"]
        raise Error, "all series must have the same number of values" unless vs.length == cats
      end
      settings ||= Settings::StackSettings.new
      settings.counts(series.length, cats)

      keep = []
      rows = series.map do |sv|
        vs = sv[:values] || sv["values"]
        name = sv[:name] || sv["name"]
        np = FFI.cstr(name)
        vp = FFI.pack_doubles(vs)
        keep << np << vp
        [[0, 8, "Q", np.to_i], [8, 8, "Q", vp.to_i]] # name @0, values @8
      end
      series_ptr = FFI.contiguous_structs(16, rows) # ccharts_stack_series = 16 B
      keep << series_ptr
      s, skeep = settings.to_ffi
      keep.concat(skeep)
      _read_string(FFI::STACK, series_ptr, series.length, width, height, s.to_ptr)
    end

    # values: a rows x cols matrix (array of equal-length arrays).
    def self.heatmap(values, width, height, settings = nil)
      raise Error, "need a non-empty value matrix" if values.empty?
      cols = values[0].length
      raise Error, "matrix columns must not be empty" if cols == 0
      values.each { |row| raise Error, "all rows must have the same number of values" unless row.length == cols }
      settings ||= Settings::HeatSettings.new

      flat = values.flatten
      flat_ptr = FFI.pack_doubles(flat)
      s, keep = settings.to_ffi
      _read_string(FFI::HEAT, flat_ptr, values.length, cols, width, height, s.to_ptr)
    end

    # series: array of {name:, samples:} hashes (per-category samples).
    def self.boxplot(series, width, height, settings = nil)
      raise Error, "need at least one category" if series.empty?
      series.each do |c|
        sm = c[:samples] || c["samples"]
        raise Error, "every category must have at least one sample" if sm.empty?
      end
      settings ||= Settings::BoxSettings.new

      keep = []
      rows = series.map do |c|
        sm = c[:samples] || c["samples"]
        name = c[:name] || c["name"]
        np = FFI.cstr(name)
        sp = FFI.pack_doubles(sm)
        keep << np << sp
        [[0, 8, "Q", np.to_i], [8, 8, "Q", sp.to_i], [16, 4, "l", sm.length]] # name@0, samples@8, n@16 (24 B)
      end
      cats_ptr = FFI.contiguous_structs(24, rows) # ccharts_box_category = 24 B
      keep << cats_ptr
      s, skeep = settings.to_ffi
      keep.concat(skeep)
      _read_string(FFI::BOX, cats_ptr, series.length, width, height, s.to_ptr)
    end

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def self._from_handle(handle)
      obj = allocate
      obj.instance_variable_set(:@handle, handle)
      ObjectSpace.define_finalizer(obj, _finalizer(handle.to_i))
      obj
    end

    def self._finalizer(addr)
      proc { FFI::DATA_FREE.call(Fiddle::Pointer.new(addr)) }
    end

    def _render(fn, width, height, settings)
      s, = settings.to_ffi
      self.class._read_string(fn, @handle, width, height, s.to_ptr)
    end

    # Call a render function and copy the returned C string, releasing it.
    def self._read_string(fn, *args)
      out = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
      out_len = Fiddle::Pointer.malloc(Fiddle::SIZEOF_SIZE_T)
      status = fn.call(*args, out, out_len)
      Ccharts.raise_on(status)
      addr = FFI.read_ptr_int(out, Fiddle::SIZEOF_VOIDP)
      len = FFI.read_ptr_int(out_len, Fiddle::SIZEOF_SIZE_T)
      s = Fiddle::Pointer.new(addr).to_s(len).force_encoding("UTF-8")
      FFI::STRING_FREE.call(Fiddle::Pointer.new(addr))
      s
    end
    private_class_method :_from_handle, :_from_data_status, :_finalizer
    private :_render
  end
end
